src/compiler/moduleSpecifiers.ts

```diff
@@ -828,51 +835,82 @@ function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: s
-                    }
+        // conditional mapping
+        for (const key of getOwnKeys(exports as MapLike<unknown>)) {
+            if (key === "default" || conditions.indexOf(key) >= 0 || isApplicableVersionedTypesKey(conditions, key)) {
+                const subTarget = (exports as MapLike<unknown>)[key];
+                const result = tryGetModuleNameFromExportsOrImports(options, targetFilePath, packageDirectory, packageName, subTarget, conditions, mode);
+                if (result) {
+                    return result;
                 }
             }
         }
     }
     return undefined;
 }
 
+function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[]): { moduleFileToTry: string } | undefined {
+    if (typeof exports === "object" && exports !== null && !Array.isArray(exports) && allKeysStartWithDot(exports as MapLike<unknown>)) { // eslint-disable-line no-null/no-null
+        // sub-mappings
+        // 3 cases:
+        // * directory mappings (legacyish, key ends with / (technically allows index/extension resolution under cjs mode))
+        // * pattern mappings (contains a *)
+        // * exact mappings (no *, does not end with /)
+        return forEach(getOwnKeys(exports as MapLike<unknown>), k => {
+            const subPackageName = getNormalizedAbsolutePath(combinePaths(packageName, k), /*currentDirectory*/ undefined);
+            const mode = endsWith(k, "/") ? MatchingMode.Directory
+                : stringContains(k, "*") ? MatchingMode.Pattern
+                : MatchingMode.Exact;
+            return tryGetModuleNameFromExportsOrImports(options, targetFilePath, packageDirectory, subPackageName, (exports as MapLike<unknown>)[k], conditions, mode);
+        });
+    }
+    return tryGetModuleNameFromExportsOrImports(options, targetFilePath, packageDirectory, packageName, exports, conditions, MatchingMode.Exact);
+}
+
+function tryGetModuleNameFromPackageJsonImports(moduleFileName: string, sourceDirectory: Path, options: CompilerOptions, host: ModuleSpecifierResolutionHost, importMode: ResolutionMode) {
+    if (!host.readFile || !getResolvePackageJsonImports(options)) {
+        return undefined;
+    }
+
+    const ancestorDirectoryWithPackageJson = getNearestAncestorDirectoryWithPackageJson(host, sourceDirectory);
+    if (!ancestorDirectoryWithPackageJson) {
+        return undefined;
+    }
+    const packageJsonPath = combinePaths(ancestorDirectoryWithPackageJson, "package.json");
+    const cachedPackageJson = host.getPackageJsonInfoCache?.()?.getPackageJsonInfo(packageJsonPath);
+    if (typeof cachedPackageJson !== "object" && cachedPackageJson !== undefined || !host.fileExists(packageJsonPath)) {
+        return undefined;
+    }
+    const packageJsonContent = cachedPackageJson?.contents.packageJsonContent || JSON.parse(host.readFile(packageJsonPath)!);

```

Comment:
I don't think this is a _new_ problem as this file already has code that looks exactly like this, but I'm somewhat certain that we should be using `readJson` for this; if the JSON is malformed, we'll throw here and then I don't know what will happen. `readJson` is what's used to get stuff into the cache, and generally speaking other raw `JSON.parse` calls all appear to at least `try/catch`.

Reply:
I see that there's another thread about this at https://github.com/microsoft/TypeScript/pull/55015#discussion_r1296222257; if we want to leave this `JSON.parse` in here given we already do that, that's fine, but overall if we're avoiding `readJson` because it calls the wrong JSON parser, but the alternatiive is to call `JSON.parse` unguarded... that feels spooky to me.